### ISSUE TITLE
[Parity 6/16] Distributed training stack - ProcessGroup, DDP, FSDP/ZeRO, DTensor, pipeline, RPC, elastic

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#215]
Part of epic #209.

## Summary

We have the NCCL transport layer (Wave 1, PR #208). What's missing is the *framework* that turns collective primitives into usable distributed training: process groups, DDP, FSDP/ZeRO, tensor/pipeline parallelism, DeviceMesh/DTensor, RPC, and elastic launch. This is the largest single gap in the audit.

### Process group abstraction
- `init_process_group(backend="nccl"/"gloo"/"mpi", world_size, rank)`, `destroy_process_group`
- `new_group`, `split_group` for sub-communicators
- Backends: NCCL (we have), Gloo (CPU / TCP / IB fallback), MPI, UCC

### Data-parallel
- `torch.nn.parallel.DistributedDataParallel` (DDP): bucketed grad reduce, gradient as bucket view, `no_sync`, `find_unused_parameters`, `static_graph`, `gradient_as_bucket_view`
- `torch.nn.parallel.DataParallel` (single-node multi-GPU — deprecated but still used)

### Sharded parallel
- FSDP (`torch.distributed.fsdp`): full/shard/hybrid strategies, mixed precision, activation offload, CPU param offload, forward/backward prefetch
- FSDP2 (`torch.distributed._composable.fsdp`)
- ZeRO-1/2/3 (optimizer / gradient / parameter sharding)

### Tensor / pipeline parallel
- `torch.distributed.pipeline` (GPipe-style), `torch.distributed.pipelining`
- `torch.distributed.tensor` (DTensor) with `DeviceMesh`, `Shard/Replicate/Partial` placement

### RPC and elasticity
- `torch.distributed.rpc`
- `torchelastic` / `torchrun` / `torch.distributed.run` (rendezvous, agent, worker)

## Current state in the repo

- `src/AiDotNet.Tensors/Engines/DirectGpu/CUDA/NcclNative.cs` + `NcclComm.cs` — bindings exist.
- Everything else missing.

## How we beat PyTorch

1. **Unified `IProcessGroup` abstraction** over NCCL + Gloo + MPI + UCC + (future) WebRTC/WebGPU mesh for browser-side sharding. PyTorch has NCCL/Gloo/MPI; we target the browser too.
2. **ZeRO-3 with algebraic pre-bucket simplification**: use `SymbolicBackwardGraph` + `BackwardPruningPass` to eliminate bucket entries for gradients that are provably zero or redundant. Expected 5–10% bucket reduction on sparse-grad models.
3. **Bucket scheduling tuned via `AutotuneCache`**: measure AllReduce latency per bucket size per cluster and persist the winner. PyTorch uses a hand-tuned default.
4. **FSDP with our native activation offload**: reuse `GpuStreamPool`, `DeferredDownload`, `DeferredScope` — zero extra infra. PyTorch reinvents the wheel.
5. **Pipeline-parallel with our graph compiler**: `LazyTensorScope` already has stage boundaries; pipelining is a scheduling pass on the existing IR, not a separate framework.
6. **DTensor placements through `TensorView` + `TensorLayoutHint`**: sharded tensors are a view over the right slice, no new storage type needed. PyTorch's DTensor wraps `Tensor` in a new subclass.
7. **Elastic launch that works on Windows** — PyTorch's `torchrun` is Linux-first. We ship a .NET-host launcher.
8. **Mixed-precision FSDP first-class**: `LayerPrecisionPolicy` already exists; FSDP inherits it automatically per shard.
9. **CPU-Gloo fallback**: unique for deploy scenarios where a GPU isn't present — training still scales on a cluster of CPU machines.

### Process groups
- [ ] `IProcessGroup` abstraction with `AllReduce`, `AllGather`, `ReduceScatter`, `Broadcast`, `Gather`, `Scatter`, `Reduce`, `Barrier`, `Send`/`Recv`, `AllToAll`
- [ ] `NcclProcessGroup` (wrap `NcclComm`)
- [ ] `GlooProcessGroup` (CPU fallback — managed TCP + optional Gloo binding)
- [ ] `MpiProcessGroup` (optional)
- [ ] `InitProcessGroup`, `NewGroup`, `SplitGroup`, `DestroyProcessGroup`

### DDP
- [ ] `DistributedDataParallel<TModel>` wrapper
- [ ] Gradient bucketing with configurable bucket size
- [ ] `no_sync` context for accumulation without communication
- [ ] Static-graph mode (mark once, skip per-iter bookkeeping)
- [ ] Uneven-input handling (`join` context)

### FSDP
- [ ] `FullyShardedDataParallel<TModel>`
- [ ] Shard / hybrid / full strategies
- [ ] Forward / backward prefetch with stream overlap
- [ ] Activation offload to CPU via `DeferredDownload`
- [ ] Param offload to CPU (optional)
- [ ] Mixed-precision policy via `LayerPrecisionPolicy`
- [ ] `FSDP.save_checkpoint` / `load_checkpoint` with sharded + consolidated modes

### ZeRO
- [ ] ZeRO-1 (optimizer state sharding)
- [ ] ZeRO-2 (optimizer + gradient sharding)
- [ ] ZeRO-3 == FSDP shard strategy

### Tensor / pipeline parallel
- [ ] `DeviceMesh` with dim-rank mapping
- [ ] `DTensor<T>` (shard/replicate/partial) as `TensorView<T>` over a mesh
- [ ] Collective ops that understand `DTensor` placements
- [ ] `Pipeline<TModel>` GPipe-style scheduler
- [ ] 1F1B / interleaved-1F1B schedules

### RPC & elastic
- [ ] `TensorRpc` — remote tensor method invocation over the process group
- [ ] Elastic launcher (.NET-native) with rendezvous over etcd / file / TCP
- [ ] Worker health checks + restart

## Acceptance criteria

- Integration tests on a simulated 2-GPU (or 2-process CPU-Gloo) setup in CI.
- DDP: ResNet-50 + ViT-Tiny train loop matches single-GPU loss curve within tolerance over N epochs.
- FSDP: 1B-param model fits in a 2×24 GB setup that would OOM under DDP.
- ZeRO-1/2/3 memory savings demonstrated.
- DTensor: explicit sharded matmul test (row-sharded × column-sharded = replicated).
- Pipeline-parallel: GPipe schedule demo with 4 stages + correct gradient flow.
- Benchmarks: AllReduce throughput on a 2-node setup vs PyTorch NCCL-backed AllReduce. Target: within 5% of PyTorch on standard sizes, clear wins on bucket scheduling via autotune.

## Implementation notes

- New module `src/AiDotNet.Tensors/Engines/Distributed/`.
- Sub-dirs: `ProcessGroups/`, `DDP/`, `FSDP/`, `Tensor/` (DTensor), `Pipeline/`, `Rpc/`, `Elastic/`.
- `IProcessGroup` implementations live next to the transport bindings (NCCL alongside `NcclComm.cs`).
- DeviceMesh integrates with `DeviceInfo.cs` / `TensorDevice.cs`.
- FSDP shard/unshard calls must be `LazyTensorScope`-aware to survive graph compile.

**Comment by ooples** [comment]
# Implementation plan — detailed

## 1. Current state (file-level)

| File | Status |
|---|---|
| `Engines/DirectGpu/CUDA/NcclNative.cs` + `NcclComm.cs` | NCCL transport ✓ |

Everything else missing: process groups, DDP, FSDP, ZeRO, DTensor, pipeline parallel, RPC, elastic launch.

## 2. Scope — full inventory

### Process groups
```
IProcessGroup { AllReduce, AllGather, ReduceScatter, Broadcast, Gather, Scatter, Reduce,
                Barrier, Send, Recv, AllToAll }
NcclProcessGroup (wrap existing NcclComm)
GlooProcessGroup (managed TCP + optional Gloo native)
MpiProcessGroup (optional, MPI.NET binding)
UccProcessGroup (bonus; heterogeneous nets)
InitProcessGroup / DestroyProcessGroup / NewGroup / SplitGroup
```

### Data-parallel
```
DistributedDataParallel<TModel> wrapper
  - Gradient bucketing (configurable bucket size; default 25MB)
  - GradientAsBucketView (zero-copy between grad and bucket)
  - NoSync context
  - StaticGraph mode
  - FindUnusedParameters
  - Uneven-input handling (join context)
DataParallel<TModel> (deprecated but shipped)
```

### FSDP / ZeRO
```
FullyShardedDataParallel<TModel>
  - Strategies: Full, Shard, Hybrid
  - ForwardPrefetch / BackwardPrefetch with stream overlap
  - ActivationOffload to CPU via DeferredDownload
  - ParamOffload to CPU (optional)
  - MixedPrecisionPolicy via LayerPrecisionPolicy
  - SaveCheckpoint / LoadCheckpoint (sharded + consolidated)
ZeRO-1 (optimizer state sharding)
ZeRO-2 (optimizer + gradient sharding)
ZeRO-3 (= FSDP shard strategy)
```

### Tensor / pipeline parallel
```
DeviceMesh with dim-rank mapping
DTensor<T> (Shard / Replicate / Partial placements) as TensorView<T> over a mesh
Collective ops that understand DTensor placements
Pipeline<TModel> GPipe-style scheduler
  - GPipe / 1F1B / Interleaved-1F1B schedules
```

### RPC & elastic
```
TensorRpc — remote tensor method invocation
Elastic launcher (.NET-native)
  - Rendezvous over etcd / file / TCP
  - Worker health checks + restart
  - torchrun-compatible CLI: `dotnet ai-tensors elastic-run`
```

**Total: ~35 surfaces.**

## 3. Win vectors

1. **Unified `IProcessGroup` over 4 backends + browser WebRTC/WebGPU mesh** — PyTorch has NCCL/Gloo/MPI; browser-side sharding unique.
2. **ZeRO-3 with algebraic pre-bucket simplification** via `SymbolicBackwardGraph` + `BackwardPruningPass` — 5–10% bucket reduction on sparse-grad models.
3. **Bucket scheduling tuned via `AutotuneCache`** per cluster (AllReduce latency × bucket size).
4. **FSDP with native activation offload** using existing `GpuStreamPool`, `DeferredDownload`, `DeferredScope` — zero new infra.
5. **Pipeline-parallel as scheduling pass on existing IR**: `LazyTensorScope` already has stage boundaries.
6. **DTensor via `TensorView` + `TensorLayoutHint`** — sharded tensors are views over right slice; no new storage type.
7. **Elastic launch that works on Windows** — PyTorch's `torchrun` is Linux-first.
8. **Mixed-precision FSDP first-class** via existing `LayerPrecisionPolicy`.
9. **CPU-Gloo fallback** — training scales on CPU-only clusters.

## 4. Per-axis beat targets

| Axis | Target |
|---|---|
| AllReduce throughput vs PyTorch NCCL on 2 nodes | ≥ parity on standard sizes |
| Bucket scheduling autotune | ≥1.05× over PyTorch fixed default |
| DDP ResNet-50 train loop | matches single-GPU loss within tolerance |
| FSDP fits 1B-param on 2×24GB | matches PyTorch scenario |
| ZeRO-1/2/3 memory savings | matches theoretical fractions |
| DTensor sharded matmul | row-sharded × col-sharded = replicated |
| Pipeline-parallel 4-stage GPipe | correct gradient flow; ≥ parity PyTorch pipeline |
| Windows elastic launch | out of box; PyTorch requires WSL |
| Backends | CPU-Gloo + NCCL required; MPI best-effort |

## 5. File inventory

**New (~90):**
- `src/AiDotNet.Tensors/Engines/Distributed/` root
- `Distributed/ProcessGroups/` — IProcessGroup + 4 backends (~20)
- `Distributed/DDP/` — DDP, bucketing, NoSync, StaticGraph, Join (~8)
- `Distributed/FSDP/` — FSDP, Shard/Hybrid/Full strategies, Prefetch, Offload (~15)
- `Distributed/Tensor/` — DeviceMesh, DTensor, placement rules (~10)
- `Distributed/Pipeline/` — GPipe, 1F1B, Interleaved schedulers (~10)
- `Distributed/Rpc/` — TensorRpc (~8)
- `Distributed/Elastic/` — launcher, rendezvous, worker agent (~15)

**Modified (~15):** `NcclComm.cs`, `DeviceInfo.cs`, `TensorDevice.cs`, `LazyTensorScope.cs`, `CompiledTrainingPlan.cs`, `ParameterBuffer.cs`, `GradientAccumulator.cs`.

**Tests (~30):** 2-process CPU-Gloo CI; multi-GPU tests skip if <2 GPUs; DDP/FSDP/ZeRO/DTensor/Pipeline integration; elastic restart.

**Benchmarks (~6):** `AllReduceBenchmarks`, `DDPResNetBenchmarks`, `FSDPLlamaBenchmarks`, `ZeroComparisonBenchmarks`, `DTensorMatMulBenchmarks`, `PipelineBenchmarks`.

## 6. Acceptance gate

1. CI runs 2-process CPU-Gloo tests green.
2. DDP: ResNet-50 convergence matches single-GPU within tolerance.
3. FSDP: 1B-param model on 2×24GB setup.
4. DTensor: sharded matmul correctness.
5. Pipeline: 4-stage GPipe gradient-flow test.
6. Windows elastic launcher starts 4 workers.
7. Mixed-precision FSDP preserves convergence.
8. XML docs with PyTorch equivalence notes.

## 7. Timeline — ~10 weeks

W1 Gloo managed TCP + IProcessGroup abstraction.
W2 NcclProcessGroup wrapping existing NcclComm.
W3 DDP bucketing + NoSync + StaticGraph.
W4 FSDP full strategy.
W5 FSDP shard + hybrid strategies + prefetch.
W6 ZeRO-1/2/3.
W7 DeviceMesh + DTensor.
W8 Pipeline scheduler (GPipe/1F1B).
W9 RPC + elastic launcher.
W10 Benchmarks + PR.

## 8. Dependencies

- **Blocks**: #216 (DistributedSampler), #218 (sharded checkpoints), #220 (distributed profiler), #224 (ZeRO-integrated optimizers).
- **Blocked by**: #214 (SavedTensorHooks for activation offload); #219 (NCCL primitives already present — no blocker).

## 9. Risks

- **CI constraints** — multi-GPU hard; CI limited to CPU-Gloo 2-process. Multi-GPU in local matrix.
- **Elastic on Windows** — process management differs; test across OS matrix.
- **DTensor placement-aware ops** — every op must respect placement; audit burden.

---

*Part of epic #209. Plan ratified 2026-04-19.*

**Comment by ooples** [comment]
## Cross-references from PR #265 (Issue #216 — Data loading)

The following data-loading items were deferred from #216 because they require infrastructure that lives in this issue. Adding here so they're tracked once the distributed stack lands:

- [ ] **Cross-machine multi-worker shared-memory transfer for `DataLoader`**: PR #265 ships a single-machine multi-worker async DataLoader using `System.Threading.Channels`. The 5–10× Windows-throughput target from #216's "How we beat PyTorch" point #1 (`MemoryMappedFiles` + pointer passing across worker processes) requires the IPC primitives this issue's `ProcessGroup` / RPC surface introduces.
- [ ] **`DistributedSampler` integration test**: PR #265 ships the sampler logic itself (deterministic stride-by-worldSize sharding, padding, per-epoch shuffle); the integration test from #216 acceptance ("`DistributedSampler` integration test with the process group from the distributed issue") needs a real `IProcessGroup` to run against.

Reference: https://github.com/ooples/AiDotNet.Tensors/pull/265

**Comment by ooples** [comment]
## FSDP reshard test inherited from #218

Issue #218's acceptance gate included:

> FSDP sharded checkpoint: save across 4 ranks, reshard to 2 ranks, weights match

That item is rolled forward to this issue rather than landing in #218's PR (which closed today via #268), because FSDP itself is part of this issue's scope. The underlying single-process `DistributedCheckpoint.Save/Load` API in `src/AiDotNet.Tensors/Serialization/Distributed/` is already in place and unit-tested; what's needed here is multi-rank wiring (ProcessGroup → FSDP shard plan → checkpoint per rank → consolidated/resharded read) plus an N→M reshard correctness test.
